### PR TITLE
Rtc alarm

### LINKED ISSUE #155: [FEATURE] Use and support alarm system on the RTC
### Short summary

Use and support alarm system on the Real Time Clock

### Motivation and user story

Adding alarm support enables features such as hibernation mode and timed system wakeups, increasing firmware flexibility and power savings. It is a foundational step for more advanced power/cycle management.

### Proposed implementation

This feature aims to add support for using the alarm system on the RTC (Real Time Clock) to unlock capabilities such as hibernation mode.

#### New Contributors: How to Approach This Issue

Please use the following steps as a guideline for your implementation:

1. **Understand the Current RTC System**
   - Review the existing `RtcManager` component and its interaction with Zephyr's RTC API.
   - Familiarize yourself with related documentation and how time setting/getting is handled.

2. **Specification & Requirements**
   - Identify which user scenarios require the use of an RTC alarm (e.g., waking the system from sleep, scheduled tasks, hibernation exit, etc.).
   - Define what "setting an alarm" means in the context of the firmware and when the alarm should trigger.

3. **Design Changes**
   - Determine what changes are required to FPP model/component interface (new commands, ports, events, or telemetry).
   - Design any data structures needed to support alarm configuration.

4. **Integrate with Zephyr RTC Alarm API**
   - Use Zephyr's driver APIs for alarm support—investigate how to set, cancel, and handle RTC alarm interrupts.

5. **Command, Event, and State Management**
   - Add commands for setting and canceling alarms.
   - Emit events for alarm successfully set, alarm trigger, alarm cancelation, and error cases.
   - Document edge cases and error handling: what happens if the RTC is not ready or an alarm fails to set?

6. **Testing**
   - Write tests or provide manual test steps showing:
     - Functionality on supported hardware.
     - Error handling for validation cases and device unavailability.
   - Update or add documentation for new feature use cases.

7. **Documentation**
   - Update relevant markdown and SDD documentation with your changes, usage instructions, and sample flows/diagrams if needed.

#### Additional Hints
- Look at the existing TIME_SET command for style and integration patterns.
- Keep code modular and follow the project's established conventions.
- If you get stuck, document what you have tried and ask for advice in your PR.
- Find the source code for the Zephyr RV3028 driver [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/rtc/rtc_rv3028.c)
- Find the application manual for the RV3028 [here](https://www.microcrystal.com/fileadmin/Media/Products/RTC/App.Manual/RV-3028-C7_App-Manual.pdf)

---

Thank you for your contribution! This task is aimed at those new to the project, so please ask questions if anything is unclear.

### Priority

_No response_

### PR DESCRIPTION
**Pull request by collinc04**
# RTC Alarm feature

## Description

Added functionality for setting and managing alarms with the zephyr API

## Related Issues/Tickets

Resolves #155 

## How Has This Been Tested?

- [ ] Unit tests
- [x] Integration tests
- [ ] Z Tests
- [x] Manual testing (describe steps)

Manual testing, Tested to make sure everything works properly. Check setting canceling and listing all in the case of alarm being present or not being present. 

## Checklist

- [x] Written detailed sdd with requirements, channels, ports, commands, telemetry defined and correctly formatted and spelled
- [x] Have written relevant integration tests and have documented them in the sdd
- [x] Have done a code review with
- [ ] Have tested this PR on every supported board with correct board definitions

## Further Notes / Considerations

Will need another feature to implement multiple alarms on a system

### CODE REVIEW

File: PROVESFlightControllerReference/test/int/rtc_test.py [hunk 1]
```diff
@@ -248,3 +248,113 @@ def test_04_sequence_cancellation_on_time_set(
...
+    fprime_test_api.send_command(f"{rtcManager}.ALARM_CANCEL")
+
+    fprime_test_api.await_event(f"{rtcManager}.AlarmTriggered", timeout=10)
+
+
+# validation test
+def test_07_rtc_alarm_set_and_cancel_validation(
+    fprime_test_api: IntegrationTestAPI, start_gds
+):
+    """Test that the alarm set and cancel commands properly validate and enforce their logic"""
```
Comment (nateinaction) [inline PROVESFlightControllerReference/test/int/rtc_test.py]:
```suggestion
def test_07_rtc_alarm_cancel_no_alarm_set(
    fprime_test_api: IntegrationTestAPI, start_gds
):
    """Test alarm cancellation when no alarm is set"""
```

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 2]
```diff
@@ -138,6 +167,126 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
...
+        this->log_ACTIVITY_HI_AlarmSet(0, t);
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::OK);
+
+    } else {
+        // if alarm IS present handle it
+        // log failure
+        // indicate no return code (user error)
+        this->log_WARNING_HI_AlarmNotSet(t, 0);
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
+    }
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: We should be able to remove this if we follow the suggestion above.

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 3]
```diff
@@ -138,6 +167,126 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
     this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::OK);
 }
 
+// Alarm manager
+void RtcManager ::ALARM_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::TimeData t) {
+    // check if alarm is already present or not if it isn't ..
+    // use this->m_dev to refer to the device]
+
+    uint16_t mask = this->curr_mask;
+    int alarmPresent = rtc_alarm_get_time(this->m_dev, 0, &mask, &this->m_alarm_time);
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]:
```suggestion
    // check if alarm is already present
    uint16_t mask = this->curr_mask;
```

minor cleanup

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 4]
```diff
@@ -138,6 +167,126 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
...
+        this->alarmTriggered_out(i);
+    }
+}
+
+void RtcManager ::ALARM_CANCEL_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, U16 ID) {
+    // check if it's present
+    uint16_t mask = this->curr_mask;
+    int rc = rtc_alarm_get_time(this->m_dev, 0, &mask, &this->m_alarm_time);
+
+    if (rc == 0 && this->alarm_set) {
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: This reads like we're checking if the alarm is set twice. What am I missing?

Either way, let's flip this logic to reduce cyclomatic complexity by

```suggestion
    // if no alarm is set, return EINVAL error
    if (rc != 0) {
        this->log_WARNING_HI_AlarmNotCanceled(ID, 22);
        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
        return;
    }
```

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 5]
```diff
@@ -138,6 +167,126 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
...
+        mask = 0;
+        rtc_alarm_set_time(this->m_dev, 0, mask, &this->m_alarm_time);
+
+        this->alarm_set = false;
+        this->log_ACTIVITY_HI_AlarmCanceled(ID);
+    } else {
+        // handle no alarm case
+        this->log_WARNING_HI_AlarmNotCanceled(ID, 0);
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
+    }
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: We can remove this logic if we adopt the suggestion above.

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.fpp [hunk 6]
```diff
@@ -14,21 +14,40 @@ module Drv {
...
 module Drv {
     @ Manages the real time clock
     passive component RtcManager {
         import Svc.Time
 
+        ### COMMANDS ###
+
+        # Alarm zephyr docs: https://docs.zephyrproject.org/latest/doxygen/html/group__rtc__interface.html#ga531c379c437a91df44a45e95063700e8
+        # When setting an alarm we should generate an ID and return it as an event
+
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.fpp]:
```suggestion
```

Remove notes

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.hpp [hunk 7]
```diff
@@ -83,6 +110,12 @@ class RtcManager final : public RtcManagerComponentBase {
     std::atomic<bool> m_console_throttled;  //!< Counter for console throttle
     const struct device* m_dev;             //!< The initialized Zephyr RTC device
     RtcHelper m_rtcHelper;                  //!< Helper for RTC operations
+
+    // rtc alarm members
+
+    U16 curr_mask;                 //!< The mask of the alarm present on hardware
+    struct rtc_time m_alarm_time;  //!< Current alarm's time settings
+    bool alarm_set;                //!< Alarm present on hardware or not
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.hpp]: I think we can remove this if suggestions above are accepted.

```suggestion
```

**Review by nateinaction — COMMENTED** [review]
This is looking awesome! Great work!

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 8]
```diff
@@ -138,6 +167,126 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
...
 
+// Alarm manager
+void RtcManager ::ALARM_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::TimeData t) {
+    // check if alarm is already present or not if it isn't ..
+    // use this->m_dev to refer to the device]
+
+    uint16_t mask = this->curr_mask;
+    int alarmPresent = rtc_alarm_get_time(this->m_dev, 0, &mask, &this->m_alarm_time);
+
+    if (alarmPresent != 0 || mask == 0) {
```
[the remark this answers is not in the document]
Comment (nateinaction) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: I made a mistake here. Instead of specifying the error numbers directly, let's import `#include <errno.h>` and then do something like:
```suggestion
    // if alarm is already set, return an error
    if (alarmPresent == 0 && mask != 0) {
        this->log_WARNING_HI_AlarmNotSet(t, EALREADY);
        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
        return;
    }
```

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 9]
```diff
@@ -138,6 +167,126 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
...
+        // assure alarm is at a future point in time
+        struct rtc_time c_time;
+        struct rtc_time a_time = this->m_alarm_time;
+        rtc_get_time(this->m_dev, &c_time);
+        int c_seconds = timeutil_timegm(rtc_time_to_tm(&c_time));
+        int a_seconds = timeutil_timegm(rtc_time_to_tm(&a_time));
+        if (a_seconds <= c_seconds) {
+            this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (collinc04) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: Added manually :)

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.hpp (hunk 7 above)
Comment (collinc04) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.hpp]: I think I am also going to make a test that will check to make sure the current alarm is ejected after one fires, I just thought about it and it seems like a good idea

**Comment by collinc04** [comment]
Last things to be updated are the changes to the alarm trigger unit test and to update the SDD, although it can probably just be merged the way it is.

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 10]
```diff
@@ -5,13 +5,30 @@
...
 // ----------------------------------------------------------------------
 // Component construction and destruction
 // ----------------------------------------------------------------------
 
-RtcManager ::RtcManager(const char* const compName) : RtcManagerComponentBase(compName), m_rtcHelper() {}
+RtcManager ::RtcManager(const char* const compName) : RtcManagerComponentBase(compName), m_rtcHelper() {
+    // alarm mask and time value struct initialization
+
+    // this->curr_mask = RTC_ALARM_TIME_MASK_MONTHDAY | RTC_ALARM_TIME_MASK_HOUR | RTC_ALARM_TIME_MASK_MINUTE;
+
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: Is this code needed to set the `this->m_curr_mask`? It looks like `RTC_ALARM_TIME_MASK_MONTHDAY | RTC_ALARM_TIME_MASK_HOUR | RTC_ALARM_TIME_MASK_MINUTE` might be built from constants. Can this be set in the header file where `U16 m_curr_mask` is declared?

File: PROVESFlightControllerReference/test/int/rtc_test.py [hunk 11]
```diff
@@ -248,3 +248,153 @@ def test_04_sequence_cancellation_on_time_set(
...
+    alarm_time = datetime.now(timezone.utc) + timedelta(seconds=5)
+    alarm_time_data = dict(
+        Year=alarm_time.year,
+        Month=alarm_time.month,
+        Day=alarm_time.day,
+        Hour=alarm_time.hour,
+        Minute=alarm_time.minute,
+        Second=alarm_time.second,
+    )
+    alarm_time_data_str = json.dumps(alarm_time_data)
```
Comment (nateinaction) [inline PROVESFlightControllerReference/test/int/rtc_test.py:389]: It looks like this isn't changing, we should remove it.

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp (hunk 10 above)
Comment (nateinaction) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: Hmmm it looks like tests are passing... is this necessary?

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 12]
```diff
@@ -138,6 +157,125 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
     this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::OK);
 }
 
+// Alarm manager
+void RtcManager ::ALARM_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::TimeData t) {
+    // check if alarm is already present or not if it isn't ..
+    // use this->m_dev to refer to the device]
+
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]:
```suggestion
```

Minor cleanup

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp (hunk 5 above)
Comment (nateinaction) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: Let's reduce cyclomatic complexity in this fn.

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp (hunk 10 above)
Comment (collinc04) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: Unnecessary, leftover from fixing EINVAL errors with board mask constraints

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 13]
```diff
@@ -138,6 +157,125 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
...
+    this->m_alarm_time.tm_year = t.get_Year() - 1900;
+
+    // assure alarm is at a future point in time
+    struct rtc_time c_time;
+    struct rtc_time a_time = this->m_alarm_time;
+    rtc_get_time(this->m_dev, &c_time);
+    int c_seconds = timeutil_timegm(rtc_time_to_tm(&c_time));
+    int a_seconds = timeutil_timegm(rtc_time_to_tm(&a_time));
+    if (a_seconds <= c_seconds) {
+        this->log_WARNING_HI_AlarmNotSet(t, 22);
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]:
```suggestion
        this->log_WARNING_HI_AlarmNotSet(t, EINVAL);
```

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp (hunk 10 above)
Comment (nateinaction) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: It looks like this still needs to be cleaned up.

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp (hunk 12 above)
Comment (nateinaction) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: It looks like this still needs to be cleaned up.

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp (hunk 5 above)
Comment (nateinaction) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: Reducing cyclomatic complexity would look like:
```cpp
    if (rc != 0 || mask == 0) {
        this->log_WARNING_HI_AlarmNotCanceled(ID, rc);
        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
        return;
    }

    // set mask to 0 to cancel alarm
    mask = 0;
    int rc = rtc_alarm_set_time(this->m_dev, 0, mask, &this->m_alarm_time);
    if (rc != 0) {
        this->log_WARNING_HI_AlarmNotCanceled(ID, rc);
        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
        return;
    }

     this->log_ACTIVITY_HI_AlarmCanceled(ID);
```

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 14]
```diff
@@ -138,6 +167,126 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
...
+
+    this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::OK);
+}
+
+void RtcManager ::ALARM_LIST_cmdHandler(FwOpcodeType opCode, U32 cmdSeq) {
+    // check if present, if so log its info.
+    uint16_t mask = this->curr_mask;
+    int rc = rtc_alarm_get_time(this->m_dev, 0, &mask, &this->m_alarm_time);
+
+    if (rc == 0) {
```
[the remark this answers is not in the document]
Comment (nateinaction) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: We should still reduce cyclomatic complexity here too.

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.fpp (hunk 6 above)
Comment (nateinaction) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.fpp]: Still need to remove notes here.

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 15]
```diff
@@ -138,6 +153,124 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
...
+    this->log_ACTIVITY_HI_AlarmTriggered(id);
+    for (int i = 0; i < getNum_alarmTriggered_OutputPorts(); i++) {
+        if (!this->isConnected_alarmTriggered_OutputPort(i)) {
+            continue;
+        }
+        this->alarmTriggered_out(i);
+    }
+    // cancel the alarm, so it won't go off repeatedly.
+    uint16_t mask = 0;
+    rtc_alarm_set_time(this->m_dev, 0, mask, &this->m_alarm_time);
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: Let's check return code.

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 16]
```diff
@@ -138,6 +153,124 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
...
+    int rc = rtc_alarm_set_time(this->m_dev, 0, this->m_curr_mask, &this->m_alarm_time);
+    if (rc != 0) {
+        // log failure
+        this->log_WARNING_HI_AlarmNotSet(t, rc);
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
+        return;
+    }
+
+    // callback to trigger upon alarm trigger
+    rtc_alarm_set_callback(this->m_dev, 0, RtcManager::static_alarm_callback_t, this);
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: Check return code

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 17]
```diff
@@ -21,8 +34,10 @@ RtcManager ::~RtcManager() {}
 
 void RtcManager ::configure(const struct device* dev) {
     this->m_dev = dev;
-}
 
+    // match to timedata this is constant after being updated here, do not change
+    rtc_alarm_get_supported_fields(this->m_dev, 0, &this->m_curr_mask);
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: Check return code. May need a new error event for this?

Comment (collinc04) [reply PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: added new event

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 18]
```diff
@@ -138,6 +157,148 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
...
+    this->m_alarm_time.tm_min = t.get_Minute();
+    this->m_alarm_time.tm_hour = t.get_Hour();
+    this->m_alarm_time.tm_mday = t.get_Day();
+    this->m_alarm_time.tm_mon = t.get_Month() - 1;
+    this->m_alarm_time.tm_year = t.get_Year() - 1900;
+
+    // assure alarm is at a future point in time
+    struct rtc_time c_time;
+    struct rtc_time a_time = this->m_alarm_time;
+    rtc_get_time(this->m_dev, &c_time);
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: Check return code

File: PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp [hunk 19]
```diff
@@ -138,6 +157,148 @@ void RtcManager ::TIME_SET_cmdHandler(FwOpcodeType opCode, U32 cmdSeq, Drv::Time
...
+        // indicates hardware error
+        this->log_WARNING_HI_AlarmHardwareError(0, rc);
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
+        return;
+    }
+
+    if (mask != 0) {
+        // set mask to 0 to cancel alarm
+        mask = 0;
+        rtc_alarm_set_time(this->m_dev, 0, mask, &this->m_alarm_time);
```
Comment (nateinaction) [inline PROVESFlightControllerReference/Components/Drv/RtcManager/RtcManager.cpp]: Check return code

**Review by nateinaction — DISMISSED** [review]
Great job!

**Review by nateinaction — APPROVED** [review]
